model_signing/hashing/file.py

```diff
@@ -0,0 +1,223 @@
+            start: The file offset to start reading from. Must be valid. Reset
+              with `set_shard`.
+            end: The file offset to start reading from. Must be stricly greater
+              than start. If past the file size, or -1, it will be trimmed.
+              Reset with `set_shard`.
+            chunk_size: The amount of file to read at once. Default is 8KB. A
+              special value of 0 signals to attempt to read everything in a
+              single call.
+            shard_size: The amount of file to read at once. Default is 8KB.
+            digest_name_override: Optional string to allow overriding the
+              `digest_name` property to support shorter, standardized names.
+        """
+        super().__init__(
+            file=file,
+            content_hasher=content_hasher,
+            chunk_size=chunk_size,
+            digest_name_override=digest_name_override,
+        )
+
+        if shard_size <= 0:
+            raise ValueError(
+                f"Shard size must be strictly positive, got {shard_size}."
+            )
+        self._shard_size = shard_size
+
+        self.set_shard(start=start, end=end)
+
+    def set_shard(self, *, start: int, end: int) -> None:
+        """Redefines the file shard to be hashed in `compute`."""
+        if start < 0:
+            raise ValueError(
+                f"File start offset must be non-negative, got {start}."
+            )
+        if end <= start:
+            raise ValueError(
+                "File end offset must be stricly higher that file start offset,"
+                f" got {start=}, {end=}."
+            )
+        read_length = end - start
+        if read_length > self._shard_size:
+            raise ValueError(
+                f"Must not read more than shard_size={self._shard_size}, got"
+                f" {read_length}."
+            )
+
+        self._start = start
+        self._end = end
+
+    @override
+    def compute(self) -> hashing.Digest:
+        self._content_hasher.reset()
+
+        with open(self._file, "rb") as f:
+            f.seek(self._start)
+            to_read = self._end - self._start
+            if self._chunk_size == 0 or self._chunk_size >= to_read:
+                data = f.read(to_read)
+                self._content_hasher.update(data)
+            else:
+                while to_read >= 0:
+                    data = f.read(min(self._chunk_size, to_read))
+                    if not data:
+                        break
+                    to_read -= len(data)
+                    self._content_hasher.update(data)
+
+        digest = self._content_hasher.compute()
+        return hashing.Digest(self.digest_name, digest.digest_value)
+
+    @override
+    @property
+    def digest_name(self) -> str:
+        if self._digest_name_override is not None:
+            return self._digest_name_override
+        return f"file-{self._content_hasher.digest_name}-{self._shard_size}"
```

Comment:
curious if we need the prepended 'file-' string or not?

Reply:
We need it, otherwise we won't be able to differentiate between an instance of the `_content_hasher` and an instance of the wrapped `FileHasher` (not the sharded one).

But we have the option to override this for hashes that we want to export in manifests